test/TritonGPU/loop-pipeline-hopper.mlir

```diff
@@ -544,3 +547,80 @@ module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-c
+    %cst_3 = arith.constant dense<0.000000e+00> : tensor<128x64xf32, #mma>
+    %cst_4 = arith.constant dense<1.000000e+00> : tensor<128x16xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #mma1}>>
+    %c1_i32 = arith.constant 1 : i32
+    %c8_i32 = arith.constant 8 : i32
+    %0 = tt.addptr %arg0, %c0_i64 : !tt.ptr<f16, 1>, i64
+    %1 = tt.addptr %arg1, %c0_i64 : !tt.ptr<f16, 1>, i64
+    %2 = tt.splat %1 : !tt.ptr<f16, 1> -> tensor<128x1x!tt.ptr<f16, 1>, #blocked1>
+    %3 = tt.addptr %2, %cst_1 : tensor<128x1x!tt.ptr<f16, 1>, #blocked1>, tensor<128x1xi32, #blocked1>
+    %4 = tt.make_range {end = 64 : i32, start = 0 : i32} : tensor<64xi32, #triton_gpu.slice<{dim = 0, parent = #blocked1}>>
+    %5 = tt.expand_dims %4 {axis = 0 : i32} : tensor<64xi32, #triton_gpu.slice<{dim = 0, parent = #blocked1}>> -> tensor<1x64xi32, #blocked1>
+    %6 = tt.broadcast %3 : tensor<128x1x!tt.ptr<f16, 1>, #blocked1> -> tensor<128x64x!tt.ptr<f16, 1>, #blocked1>
+    %7 = tt.broadcast %5 : tensor<1x64xi32, #blocked1> -> tensor<128x64xi32, #blocked1>
+    %8 = tt.addptr %6, %7 : tensor<128x64x!tt.ptr<f16, 1>, #blocked1>, tensor<128x64xi32, #blocked1>
+    %9 = tt.load %8 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<128x64xf16, #blocked1>
+    %10 = tt.splat %0 : !tt.ptr<f16, 1> -> tensor<1x16x!tt.ptr<f16, 1>, #blocked>
+    %11 = tt.addptr %10, %cst_0 : tensor<1x16x!tt.ptr<f16, 1>, #blocked>, tensor<1x16xi32, #blocked>
+    %12 = tt.make_range {end = 64 : i32, start = 0 : i32} : tensor<64xi32, #triton_gpu.slice<{dim = 1, parent = #blocked}>>
+    %13 = tt.expand_dims %12 {axis = 1 : i32} : tensor<64xi32, #triton_gpu.slice<{dim = 1, parent = #blocked}>> -> tensor<64x1xi32, #blocked>
+    %14 = tt.broadcast %11 : tensor<1x16x!tt.ptr<f16, 1>, #blocked> -> tensor<64x16x!tt.ptr<f16, 1>, #blocked>
+    %15 = tt.broadcast %13 : tensor<64x1xi32, #blocked> -> tensor<64x16xi32, #blocked>
+    %16 = tt.addptr %14, %15 : tensor<64x16x!tt.ptr<f16, 1>, #blocked>, tensor<64x16xi32, #blocked>
+    %18 = tt.load %16 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<64x16xf16, #blocked>
+    %19 = triton_gpu.local_alloc %9 : (tensor<128x64xf16, #blocked1>) -> !tt.memdesc<128x64xf16, #shared>
+    %20 = triton_gpu.local_alloc %18 : (tensor<64x16xf16, #blocked>) -> !tt.memdesc<64x16xf16, #shared1>
+    // CHECK:          %[[LOOP:[^ :]+]]{{.*}} scf.for {{.*}} iter_args(%[[PREV_DOT2:[^ ]+]]
+    // CHECK:            %[[DOT0:.+]] = triton_nvidia_gpu.dot_async
+    // CHECK:            %[[DOT1:.+]] = triton_nvidia_gpu.dot_async
+    // CHECK-NEXT:       triton_nvidia_gpu.dot_wait
+    // CHECK-DAG-SAME:     %[[DOT0]]
+    // CHECK-DAG-SAME:     %[[DOT1]]
+    // CHECK-DAG-SAME:     %[[PREV_DOT2]]
+    // CHECK-SAME:         {pendings = 0 : i32}
+    // CHECK:            %[[DOT2:.+]] = triton_nvidia_gpu.dot_async
+    // CHECK:            %[[WAIT:[^ :]+]]:{{[0-9]+}} = triton_nvidia_gpu.dot_wait
```

Comment:
Is this wait needed? I thought part of the logic was to remove this kind of wait?

Reply:
The testcase is actually set up specifically so that you can't remove this wait.  dot1 is synchronous, but dot0 is not.  You can only get rid of the wait at the end of the loop if the first or the last dot in the loop is synchronous.

Reply:
but the wait has pending = 2, so it will never do anything since we have the code sequence:
```
dot_async
dot_async
wait 0
dot_async
wait 2 // There is max 1 dot_async in flight here (the one right above)
```
is that expected? It feels like something is off in how the pendings calculation is done.

Reply:
so the last dot has a register use after the sync dot so the IR is technically correct although the second wait is not doing anything. What would happen if the `addf` was before the sync dot?

Reply:
> What would happen if the addf was before the sync dot?

I'm pretty sure we'd still have a wait at the end of the loop -- we only get rid of it if there's a sync dot at the beginning or end of the loop.

This PR wasn't trying to get rid of nop waits, but rather to allow more dots to be async.  My thought is that nop waits aren't expensive so it's nbd?  But if you want I can definitely get rid of the nop wait in this case.

Reply:
> > What would happen if the addf was before the sync dot?
> 
> I'm pretty sure we'd still have a wait at the end of the loop -- we only get rid of it if there's a sync dot at the beginning or end of the loop.

Might be worth changing the test to test this case as it seems like the interesting one where the IR would be wrong if we don't generate a wait 0 

> This PR wasn't trying to get rid of nop waits, but rather to allow more dots to be async. My thought is that nop waits aren't expensive so it's nbd? But if you want I can definitely get rid of the nop wait in this case.

Yes that's fine a long as the logic is correct for the case where we do need to wait. I guess I'm thrown off by pending = 2 but if for the cases mentioned above we generate the right wait then it's fine.

Reply:
> Might be worth changing the test to test this case as it seems like the interesting one [is?] where the IR would be wrong if we don't generate a wait 0

The IR in this testcase would be wrong if we didn't generate a wait 0, no?  dot1 is used immediately afterwards by an addf, so it needs to be synchronized.

I am not sure what you're asking me to change.  Is it that you want me to move the addf above the sync dot?  That would make the dot not sync anymore -- it's specifically that use of the dot that makes it synchronous.  Then we would no longer be testing the case where the wait 0 is required.

Reply:
> The IR in this testcase would be wrong if we didn't generate a wait 0, no? dot1 is used immediately afterwards by an addf, so it needs to be synchronized.

I meant the second wait.

> I am not sure what you're asking me to change. Is it that you want me to move the addf above the sync dot? That would make the dot not sync anymore -- it's specifically that use of the dot that makes it synchronous. Then we would no longer be testing the case where the wait 0 is required.

Right I meant moving `%prev_dot2.1 = arith.addf %prev_dot2, %prev_dot2 : tensor<128x64xf32, #mma>` between the two dots to make sure the second wait is a wait 0. Unless this is already tested. 
Anyway I was mostly confused about how we end up with a wait 2 here but if you are saying this is what you were expecting this is fine.